### DISCUSSION TITLE
Is it possible for nx affected to check for file changes instead?

### DISCUSSION THREAD

**Discussion by anthonyma94** [nrwl/nx#13852] [General]
We've been using nx at work for about a year now, and I'd like to make some improvements to our CI/CD. Most notably, we are currently manually feeding tags into our pipeline to tell it what's been changed, and I'd like to have that be done automatically. 

`nx affected` seems like a great option, but it seems to track changes on a project level. This doesn't really work for us: we have several microservices that are only reliant on 1 or 2 files in a specific project. I'll give you a condensed look of our project:
```
- apps
  - api
  - jobs
- libs
  - db-services
    - UserService.ts
    - IOTService.ts
    - DeviceService.ts
    - ...
```
Our `jobs` app only needs access to `IOTService.ts`, but since that is in the `db-services` project, every time any other service is updated in the project, `nx affected` will build `jobs` as well. Is there a way to have the command check only the files affected? If not, do I need to restructure my code? How can I make it more efficient?

**Comment by veloware** [comment]
one option would be to split your `db-services` further, taking your current example - 

```
- apps
  - api
  - jobs
- libs
   - IOT (directory)
      - db-service.ts (lib)
   - user (directory)
      - db-service.ts (lib)
   - device (directory)
      - db-service.ts (lib)
   - shared (directory) (to house shared libs, across IOT, user, device domains etc)
      - db-service.ts (lib)
```

now `jobs` can consume only the `IOT/db-service.ts` and will get built only when that service has been affected. You will of course have to rebuild the other apps if they also rely on that service, as before.

More reading - https://nx.dev/more-concepts/grouping-libraries

**Reply by anthonyma94** [reply]
My problem with this is that the service is just a single file. I have about 30 different services, it's unrealistic to generate 30 projects with a single file in each.

**Reply by veloware** [reply]
I thought that too but if you follow the 'mental model' they describe in their docs and use their generators when creating libs it's actually quite painless and will solve for what you're trying to achieve. I've only started a project and already have 24 libs, all starting to organically shape the architecture behind my app 

https://nx.dev/more-concepts/applications-and-libraries#mental-model

And if you think about your statement, around having a separate lib per service, that's actually quite reasonable to do

**Reply by anthonyma94** [reply]
You'd also be generating 30 project's worth of config files. A new node library comes with `tsconfig.json`, `tsconfig.lib.json`, `.eslintrc`, and `.babelrc` at minimum. If you want testing, add `jest.config.ts` and `tsconfig.spec.json`. That's 6 additional files in your repo, for the 1 file you're adding. Not to mention our repo is designed to have a `project.json` for each project, and that our configs are modified from the default. That's just not maintanable.

**Reply by veloware** [reply]
This is all part of the trade off though, right? I also agree some of the files generated feel a bit over the top, but still, you want the ability to breakdown your application into specific dependencies to deploy independently but aren't expecting to manage some sort of config for each of those dependencies / libraries. Many of those config files inherit from a base, so common changes are still easily managed, e.g. tsconfig.base.json and the generators do a great job of creating, deleting, editing or moving libs

 The docs layout the recommended way to structure your app and libs to make the best use of NX features. It's a hard sell to question why things aren't working as you expect when you don't follow the recommended structure etc
